### PR TITLE
Remove shared_ptr

### PR DESCRIPTION
**Pull request by davidbeckingsale**
Unnecessary because we don't have shared ownership of objects, and `shared_ptr` adds overhead.

### CODE REVIEW

File: examples/CMakeLists.txt [hunk 1]
```diff
@@ -52,6 +52,12 @@ if (ENABLE_STATISTICS)
     OUTPUT_DIR ${UMPIRE_EXAMPLES_OUTPUT_DIR})
 endif()
 
+#blt_add_executable(
```
Comment (mcfadden8) [inline examples/CMakeLists.txt]: What is sandbox and how does it relate to shared_ptr changes?

File: src/umpire/op/MemoryOperationRegistry.cpp [hunk 2]
```diff
@@ -195,8 +195,8 @@ MemoryOperationRegistry::registerOperation(
 std::shared_ptr<umpire::op::MemoryOperation>
 MemoryOperationRegistry::find(
     const std::string& name,
-    std::shared_ptr<strategy::AllocationStrategy>& src_allocator,
-    std::shared_ptr<strategy::AllocationStrategy>& dst_allocator)
+    strategy::AllocationStrategy*& src_allocator,
```
Comment (jdahm) [inline src/umpire/op/MemoryOperationRegistry.cpp]: Should we remove the reference to the pointer?

File: src/umpire/resource/CudaConstantMemoryResourceFactory.cpp [hunk 3]
```diff
@@ -32,7 +32,7 @@ CudaConstantMemoryResourceFactory::isValidMemoryResourceFor(const std::string& n
   }
 }
 
-std::shared_ptr<MemoryResource>
+resource::MemoryResource*
```
Comment (jdahm) [inline src/umpire/resource/CudaConstantMemoryResourceFactory.cpp]: Do we need `resource::` here for this to work?

**Comment by jdahm** [comment]
Looks great 👍 !

**Comment by mcfadden8** [comment]
@davidbeckingsale, what is the relative performance improvement of this?

**Comment by davidbeckingsale** [comment]
Good question. Going to benchmark it today, and ask @robinson96 to test it out too...

**Comment by davidbeckingsale** [comment]
This PR:
```
Run on (4 X 2900 MHz CPU s)
2019-03-21 09:33:02
--------------------------------------------------------------------------
Benchmark                                   Time           CPU Iterations
--------------------------------------------------------------------------
benchmark_allocator_create                  2 ns          2 ns  340569627
benchmark_allocator_create_destroy          2 ns          2 ns  338362038
benchmark_allocator_get                    52 ns         50 ns   10000000
```

`develop`
```
Run on (4 X 2900 MHz CPU s)
2019-03-21 09:30:21
--------------------------------------------------------------------------
Benchmark                                   Time           CPU Iterations
--------------------------------------------------------------------------
benchmark_allocator_create                  2 ns          2 ns  339904827
benchmark_allocator_create_destroy          2 ns          2 ns  342955690
benchmark_allocator_get                    71 ns         71 ns    9512291
```

So there is some improvement in the 'get' benchmark which tests the whole pipeline of getting an Allocator from the ResourceManager. It's hard to spot in the allocator creation. I'll commit these benchmarks to this branch.

**Comment by davidbeckingsale** [comment]
@mcfadden8 @jdahm - can I get a review?

File: examples/CMakeLists.txt (hunk 1 above)
Comment (mcfadden8) [reply examples/CMakeLists.txt]: Again, what is the sandbox stuff @davidbeckingsale ?

File: src/umpire/op/MemoryOperationRegistry.cpp (hunk 2 above)
Comment (jdahm) [reply src/umpire/op/MemoryOperationRegistry.cpp]: This was addressed 👍

File: src/umpire/ResourceManager.hpp [hunk 4]
```diff
@@ -241,9 +241,9 @@ class ResourceManager {
     ResourceManager (const ResourceManager&) = delete;
     ResourceManager& operator= (const ResourceManager&) = delete;
 
-    std::shared_ptr<strategy::AllocationStrategy>& findAllocatorForPointer(void* ptr);
-    std::shared_ptr<strategy::AllocationStrategy>& findAllocatorForId(int id);
-    std::shared_ptr<strategy::AllocationStrategy>& getAllocationStrategy(const std::string& name);
+    strategy::AllocationStrategy*& findAllocatorForPointer(void* ptr);
```
Comment (jdahm) [inline src/umpire/ResourceManager.hpp]: More references.

File: src/umpire/ResourceManager.cpp [hunk 5]
```diff
@@ -177,7 +177,7 @@ ResourceManager::initialize()
   UMPIRE_LOG(Debug, "() leaving");
 }
 
-std::shared_ptr<strategy::AllocationStrategy>&
+strategy::AllocationStrategy*&
```
Comment (mcfadden8) [inline src/umpire/ResourceManager.cpp]: @davidbeckingsale - Why does this (and some others) return a reference to a pointer instead of just a pointer?

**Pull request merged**